packages/kilt-did/src/types.ts

```diff
@@ -0,0 +1,104 @@
+import type { IIdentity, SubmittableExtrinsic } from '@kiltprotocol/types'
+import { u64 } from '@polkadot/types'
+import { AnyNumber } from '@polkadot/types/types'
+import { KeyId } from './types.chain'
+
+export type KeypairType = string
+
+export interface IPublicKey {
+  publicKey: Uint8Array
+  type: KeypairType
+  // do we need the identifier?
+}
+
+export type IKeyPair = IPublicKey
+
+export type ISigningKey = IPublicKey
+export type IEncryptionKey = IPublicKey
+
+export interface ISigningKeyPair extends ISigningKey {
+  sign: (message: string | Uint8Array) => Uint8Array
+}
+
+export interface IEncryptionKeyPair extends IEncryptionKey {
+  decrypt: (
+    message: Uint8Array,
+    additionalData: Record<string, unknown>
+  ) => Uint8Array | null
+}
+
+// TODO: should this use the same keys as the chain type?
+export interface KeySet {
+  authentication: ISigningKeyPair
+  encryption: IEncryptionKeyPair
+  attestation?: ISigningKeyPair
+  delegation?: ISigningKeyPair
+}
+
+export interface PublicKeySet {
```

Comment:
Same, something like `ActivePublicKeySet`.

Reply:
Could also be named `KeyRolesMap` or something similar, as it maps roles/functions/verification relationships to keys? 
Honestly, I am still not super happy with the way keys are managed here. Not only can we have multiple encryption keys, which this does not represent. It also doesn't really represent the fact very well that on chain we only have a vector of keys, with references to them identifying the roles these keys are meant to be used in.
So if you have any ideas for alternatives, let me know.

Reply:
I guess one alternative would be to manage keys as an Array, mirroring how they are stored on chain, and annotate them with their respective roles by adding an additional "role" property where applicable: 
```js
[ 
  { id: '0x...', publicKeyHex: '0x...', type: 'x25519', role: 'KeyAgreement', includedAt: 110},
  { id: '0x...', publicKeyHex: '0x...', type: 'sr25519', role: 'Authentication', includedAt: 85},
  { id: '0x...', publicKeyHex: '0x...', type: 'sr25519', includedAt: 85}
]
```

Reply:
I don't think storing keys in an array would be the most efficient solution. I think it is probably more clear in which direction to go after yesterday's discussion, and my point was more about the naming of this interface than anything else really.

Reply:
I called the interface `PublicKeyRoleAssignment`, hope that works for you